Add the GitProvider Client interfaces and types

luxas opened the pull request:
As per the proposal up at: https://toolkit.fluxcd.io/proposals/go-git-providers/

This PR checks in the common interfaces, types and helper functions (like parsing URLs).
I will do one more PR for setting up the scaffolding for the specific providers.
Then, future PRs, by authors of e.g. the code in `weaveworks/go-git-providers` or `fluxcd/pkg/git`, can add the actual provider-specific code.

Please review.
The API has changed a bit since I wrote the proposal doc, but I'll reflect these changes in the doc a bit later.

I put my name in the maintainers list for the moment, when I help bootstrap and design this repo.

This PR has 100% unit test coverage, hope we can keep things that way :smile:

cc @stefanprodan @hiddeco @squaremo @martina-if @dinosk

stefanprodan commented on gitprovider.go:
```diff
@@ -0,0 +1,115 @@
+	// ErrURLUnsupportedScheme is returned if an URL without the HTTPS scheme is parsed
+	ErrURLUnsupportedScheme = errors.New("unsupported URL scheme, only HTTPS supported")
+	// ErrURLUnsupportedParts is returned if an URL with fragment, query values and/or user information is parsed
+	ErrURLUnsupportedParts = errors.New("organization or repository URLs cannot have fragments, query values nor user information")
```
I think we should either create dedicated errors for organization and repository or make the current messages generic as in: 

```
ErrURLUnsupportedParts = errors.New("URL cannot have fragments, query values nor user information")
```

Doesn't make sense to me to say `URLs` when in fact you validate a single URL, having `organization or repository URLs` confuses the end-user as it doesn't tell it what URL is actually wrong. I think the error should contain the URL string.

luxas replied:
Thanks. Changed to your suggestion, that's better. The actual URL which is wrong will be added as context at runtime, but this can error can be used for `errors.Is(err, ErrURLUnsupportedParts)`

stefanprodan replied:
Great! thanks

twelho commented on gitprovider.go:
```diff
@@ -0,0 +1,115 @@
+
+// Creator is an interface which all objects that can be updated
+// (using the Client) should implement
+type Updator interface {
```
[Updator vs updater](https://english.stackexchange.com/questions/16161/one-who-creates-is-a-creator-what-is-one-who-updates), updater seems to be the more correct term of the two. I'd also argue that this should be `Updatable`, since it describes the capability of receiving an update.

twelho commented on gitprovider.go:
```diff
@@ -0,0 +1,115 @@
+	Default()
+}
+
+// Creator is an interface which all objects that can be updated
```
Typo: `Creator`

twelho commented on gitprovider.go:
```diff
@@ -0,0 +1,115 @@
+
+// Deletor is an interface which all objects that can be deleted
+// (using the Client) should implement
+type Deletor interface {
```
Same as above, `Deletable` suits this better IMO, the ability to be deleted.

twelho commented on gitprovider.go:
```diff
@@ -0,0 +1,115 @@
+	return ih.Internal
+}
+
+func boolVar(b bool) *bool {
```
I'd keep this private but move it to `util.go` or similar (and add a one-line comment). This helps in keeping the interface definition files clean.

twelho commented on options.go:
```diff
@@ -0,0 +1,70 @@
+package gitprovider
+
+// RepositoryCreateOptionsFunc is a function mutating the given RepositoryCreateOptions argument
+// This function is passed a variable amount into RepositoriesClient.Create()
```
`// A variable amount of these functions can be passed into RepositoriesClient.Create()`

twelho commented on options.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+// MakeRepositoryCreateOptions returns a RepositoryCreateOptions based off the mutator functions
+// given to e.g. RepositoriesClient.Create(). The returned validation error might be ignored in the
```
`might` => `may`

twelho commented on repositoryref.go:
```diff
@@ -0,0 +1,258 @@
+	// String returns the HTTPS URL
+	fmt.Stringer
+
+	// GetDomain returns the URL-domain for the Git provider backend, e.g. gitlab.com or version.aalto.fi
```
Should we replace `version.aalto.fi` with some imaginary/common alternative GitLab provider?

twelho commented on repositoryref.go:
```diff
@@ -0,0 +1,258 @@
+
+	// GetDomain returns the URL-domain for the Git provider backend, e.g. gitlab.com or version.aalto.fi
+	GetDomain() string
+	// GetOrganization returns the top-level organization, i.e. "weaveworks" or "kubernetes-sigs"
```
Since this is in the `fluxcd` org/repo, I'd replace `weaveworks` with `fluxcd` (or another alternative).

twelho commented on repositoryref.go:
```diff
@@ -0,0 +1,258 @@
+// ParseRepositoryURL parses a HTTPS or SSH clone URL into a RepositoryRef object
+func ParseRepositoryURL(r string) (RepositoryRef, error) {
+	// First, parse the URL as an organization
+	orgInfo, err := parseOrganizationURL(r)
```
Call this `orgInfoPtr` to avoid confusion with the above function.

twelho commented on repositoryref.go:
```diff
@@ -0,0 +1,258 @@
+		return nil, nil, ErrURLUnsupportedScheme
+	}
+	// Don't allow any extra things in the URL, in order to be able to do a successful
+	// round-trip or parsing the URL and encoding it back to a string
```
`or` => `of`?

stefanprodan commented:
@luxas does the API accommodate personal repos outside of an organisation? This is a requirement for `tk boostrap` as we don't want to force users into creating an org.

stefanprodan commented on repositoryref.go:
```diff
@@ -0,0 +1,258 @@
+func GetCloneURL(rs RepositoryRef, transport TransportType) string {
+	switch transport {
+	case TransportTypeHTTPS:
+		return fmt.Sprintf("%s.git", rs.String())
```
What if the URL contains `.git`, we should not append it that case.

luxas replied:
Thanks for catching that.
I'm now excluding any possible `.git` suffix when parsing the URL to make `rs.String()` always return the HTTPS URL for the web page (without the .git suffix to make it safe to append)

stefanprodan commented on .github/workflows/golangci.yaml:
```diff
@@ -0,0 +1,14 @@
+        uses: golangci/golangci-lint-action@v1
+        with:
+          # Required: the version of golangci-lint is required and must be specified without patch version: we always use the latest patch version.
+          version: v1.28
```
New line here please

luxas commented:
> does the API accommodate personal repos outside of an organisation? This is a requirement for tk boostrap as we don't want to force users into creating an org.

It'll work the same for any user too :+1:
I don't know what would be a better word than "organization" to represent "user login" or "organization" (in Github) or "group/sub-group" (in Gitlab).

dinosk commented on client.go:
```diff
@@ -0,0 +1,168 @@
+
+// RepositoryClient operates on a given/specific repository
+type RepositoryClient interface {
+	// TeamAccess gives access to what teams have access to this specific repository
```
if the `RepositoryTeamAccessClient` can also create/delete teams on the control list, maybe the comment can be rephrased to something like `returns a client for operating on the teams that have access to this specific repository` ?

luxas replied:
thx, fixed

dinosk commented:
This looks great @luxas :+1:

stefanprodan commented:
> It'll work the same for any user too

It's not the same thing, for personal repos the organisation must be set to string empty, otherwise GitHub API will error out with org not found. https://github.com/fluxcd/pkg/blob/master/git/provider_github.go#L63-L66

stefanprodan commented:
@luxas we use go tidy, fmt and vet across fluxcd projects, can you please remove the lint job or fix it?

luxas commented:
> we use go tidy, fmt and vet across fluxcd projects, can you please remove the lint job or fix it?

Fixed. I think having lint is really useful.

> It's not the same thing, for personal repos the organisation must be set to string empty, otherwise GitHub API will error out with org not found

Oh, I had missed that, thanks for notifying. The API is now split into `IdentityRef`, `UserRef` and `OrganizationRef`, instead of only `OrganizationRef`. It's now explicit in the code what is expecting an organization and what is expecting an user account.

This should be ready to merge. It has 100% unit test coverage.
